File: code/algo/getoffer/go/3/arr_sort_base_arr2.go
package main

import "sort"

type zset map[int]int

func (zs *zset) keys() []int {
	res := []int{}
	for key := range *zs {
		res = append(res, key)
	}
	sort.Ints(res)
	return res
}

func relativeSortArray(arr1 []int, arr2 []int) []int {
	zs := zset{}
	for _, val := range arr1 {
		zs[val]++
	}

	res := make([]int, 0, len(arr1))
	for _, val := range arr2 {
		for i := 0; i < zs[val]; i++ {
			res = append(res, val)
		}
		delete(zs, val)
	}
	for _, key := range zs.keys() {
		for i := 0; i < zs[key]; i++ {
			res = append(res, key)
		}
	}
	return res
}

File: code/algo/getoffer/c++/3/arr_sort_base_arr2.cpp
#include <algorithm>
#include <unordered_map>
#include <vector>

using namespace std;

class ZSet {
public:
  void insert(int key) { data[key]++; }

  void erase(int key) { data.erase(key); }

  vector<int> keys() const {
    vector<int> res;
    for (const auto &[key, _] : data) {
      res.push_back(key);
    }
    sort(res.begin(), res.end());
    return res;
  }

  int count(int key) const {
    auto it = data.find(key);
    return it != data.end() ? it->second : 0;
  }

private:
  unordered_map<int, int> data;
};

vector<int> relativeSortArray(vector<int> &arr1, const vector<int> &arr2) {
  ZSet zs;
  for (int val : arr1) {
    zs.insert(val);
  }

  vector<int> res;
  for (int val : arr2) {
    int count = zs.count(val);
    res.insert(res.end(), count, val);
    zs.erase(val);
  }

  vector<int> remainingKeys = zs.keys();
  for (int key : remainingKeys) {
    int count = zs.count(key);
    res.insert(res.end(), count, key);
  }

  return res;
}
